clientconn.go

```diff
@@ -1331,19 +1331,32 @@ func (ac *addrConn) createTransport(addr resolver.Address, copts transport.Conne
 	case <-prefaceReceived.Done():
 		// We got the preface - huzzah! things are good.
 		ac.mu.Lock()
-		defer ac.mu.Unlock()
-		defer prefaceReceived.Fire()
 		if connClosed.HasFired() {
 			// onClose called first; go idle but do nothing else.
 			if ac.state != connectivity.Shutdown {
 				ac.updateConnectivityState(connectivity.Idle, nil)
 			}
+			ac.mu.Unlock()
 			return nil
 		}
+		if ac.state == connectivity.Shutdown {
+			// This can happen if the subConn was removed while in `Connecting`
+			// state. tearDown() would have set the state to `Shutdown`, but
+			// would not have closed the transport since ac.transport would not
+			// been set at that point.
+			err := fmt.Errorf("subConn closed before receiving preface")
```

Comment:
Actually this error text is potentially unclear, too. It makes it sound like the connection was lost before we received the preface.  Why not just say something like "subConn shutdown by balancer"?

Reply:
Changed the text for this error.